// framed-dal/lib/frameDAO.js
var dal = require('./base/mongoDAO');



module.exports = function(options){

    /*DAO for Frame objects*/
    function FrameDAO(){}

    options = options || {};

    //ensure default collection name
    if(!options.collection){
        options.collection ='frames';
    }

    //Init default validator(s)
    options.validators = options.validators || [];
    options.validators.push(function (item) {
        var v = new Validator();
        var measurementSchema = {

            "id": "/Measurement",
            "type": "object",
            "properties": {

                "value": {"type": "number"},
                "unit": {"type": "string"},
                "auto": {"type": "boolean"}
            }
        };

        var attribSchema = {

            "id": "/Attribute",
            "type": "object",
            "properties": {
                "name": {"type": "string", "required": true},
                "value": {"type": "string"}
            }

        }

        var frameSchema = {
            "id": "/Frame",
            "type": "object",
            "properties": {
                "frames": {
                    "type": "array",
                    "items": {"$ref": "/Frame"}
                },
                "classes": {
                    "type": "array",
                    "items": {"type": "string"}
                },
                "attributes": {
                    "type": "array",
                    "items": {"$ref": "/Attribute"}
                },
                "name": {"type": "string"},
                "version": {"type": "string", "required": true},
                "width": {"$ref": "/Measurement"},
                "height": {"$ref": "/Measurement"},
                "topMargin": {"$ref": "/Measurement"},
                "rightMargin": {"$ref": "/Measurement"},
                "bottomMargin": {"$ref": "/Measurement"},
                "leftMargin": {"$ref": "/Measurement"}
            }
        };

        v.addSchema(measurementSchema, '/Measurement');
        v.addSchema(attribSchema, '/Attribute');
        v.addSchema(frameSchema, '/Frame');

        var result = v.validate(item, frameSchema);
        return result.errors.length < 1;
    });


    //Init default object creation source
    options.getSeed = function () {
        return {
            name: 'New Frame',
            version: '1.0.0',
            classes: [],
            attributes: [],
            width: {value: 0, unit: 'px', auto: true},
            height: {value: 0, unit: 'px', auto: true},
            topMargin: {value: 0, unit: 'px', auto: true},
            rightMargin: {value: 0, unit: 'px', auto: true},
            bottomMargin: {value: 0, unit: 'px', auto: true},
            leftMargin: {value: 0, unit: 'px', auto: true}
        };
    };

    dal.mongoDAO(options, FrameDAO.prototype);

    return new FrameDAO();
};
